src/tfs/tlog.c

```diff
@@ -477,7 +477,11 @@ static void run_flush_completions(log tl, status s)
     if (tl->flush_completions) {
         status_handler sh;
         vector_foreach(tl->flush_completions, sh)
+#ifdef KERNEL
+            async_apply_status_handler(sh, s);
```

Comment:
It looks like log_extend_link would sometimes be run async this way, which modifies the old extension's staging buffer outside of any lock in that situation, which maybe is okay? It's implied that this closure would be the only reference so it might be fine.

My other concern is guaranteeing this completion is run in a shutdown situation where this might be left on an async queue and thus missing an fs metadata write if the filesystem sync completes first.

Reply:
> It looks like log_extend_link would sometimes be run async this way, which modifies the old extension's staging buffer outside of any lock in that situation, which maybe is okay? It's implied that this closure would be the only reference so it might be fine.

Yes, I think it's ok for log_extend_link() to be called without the TFS lock, because it's the last action being made on a log extension before closing it, so no other parts of the code should access that extension.

> My other concern is guaranteeing this completion is run in a shutdown situation where this might be left on an async queue and thus missing an fs metadata write if the filesystem sync completes first.

During shutdown, filesystem_flush() is called, which first flushes the TFS log (thus writing any pending filesystem metadata), and then on flush completion syncs the page cache volume, which takes care of writing any pending file data. Only after this last action is completed, vm_exit() is called. So it seems we cannot miss a metadata write.

Reply:
Yeah, I was thinking of the case where a rebuild happened at the same time as shutdown, which could cause the log_extend_link to run off the async queue, which would generate new i/o and whether or not the shutdown sync could race by that. It looks like the sync would be dependent on the log_extend_link to finish (which leads to the rebuild finishing), and then the sync would run.  Thanks!